### PR TITLE
[1/?] Add pre-commit config.

### PR DESCRIPTION
**Pull request by AllSeeingEyeTolledEweSew**
[pre-commit](https://pre-commit.com) is a CI ecosystem that can be run locally.

It's designed to organize pre-commit git hooks, so it's good for quick static analysis checkers and autoformatters, though not really good for automatic builds and tests. It's widely used in the open source python world today. I've been using it for a few months and I like it a lot.

This PR:
* adds pre-commit config
* runs pre-commit in GH actions
* enables some trailing-whitespace auto-formatters and linters, with some trivial changes in files
* moves `flake8` invocation from travis to pre-commit

This is really an initial support for later PRs. I will make heavy use of this in later PRs which expand the python tests.

I think these commits are pretty reasonable, but I can squash if you like.

### CODE REVIEW

**Comment by AllSeeingEyeTolledEweSew** [comment]
The `exclude: ` config is just to prevent bloating this PR.

**Comment by arvidn** [comment]
is running these checks still optional? opt-in? or is does this make `flake8` a requirement to commit anything?

I assume it's optional.

**Review by arvidn — CHANGES_REQUESTED** [review]
I think there are certain files that should be exempt, that are included verbatim in the repository. Is it easy to add exemptions?
Specifically, these files:

* anything under docs/hunspell/
* anything under ed25519/
* include/libtorrent/aux_/route.h
* test/*.xml
* test/ssl/*.pem

Also, travis isn't running anymore, I will remove the .travis.yml.
Also, I don't think docs/make_thumb.sh is in use anymore, I will look at removing that as well.

How does one install and run "pre-commit"? Would you mind describing this in CONTRIBUTING.rst

File: bindings/python/setup.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 import platform
 import sys
 import sysconfig
-import tempfile
+# import tempfile
```
[the remark this answers is not in the document]
Comment (AllSeeingEyeTolledEweSew) [reply bindings/python/setup.py:12]: I did this in #5902 which was targeted at more thorough cleanup of `setup.py`

**Comment by AllSeeingEyeTolledEweSew** [comment]
> is running these checks still optional? opt-in? or is does this make `flake8` a requirement to commit anything?
> 
> I assume it's optional.

It's optional to run the checks locally. The checks are still functionally required since they're run in CI.

pre-commit is git-aware. when running locally, it won't run `flake8` at all if you didn't change python code. (all checks are always run in CI, in this setup)

the goal is that all contributors should be motivated to always run pre-commit (as a git hook), and if there are no relevant changes it'll do nothing, and if there are relevant changes it'll save time by showing failures early rather than waiting for CI to show them.

> I think there are certain files that should be exempt, that are included verbatim in the repository. Is it easy to add exemptions?

It's easy to add, I'll add those.
 
> How does one install and run "pre-commit"? Would you mind describing this in CONTRIBUTING.rst

```bash
$ pip install pre-commit
$ pre-commit install  # if you want to run it as a git hook
```

I'll add that to CONTRIBUTING.rst

**Comment by AllSeeingEyeTolledEweSew** [comment]
I think the CI failure can't be mine (?)

**Comment by arvidn** [comment]
sorry. I continued the conversation in the `[2/?]` PR.

I think the only things I'm still not sold on is that the value of running the pre-commit checks on 4 different virtual machines carries its weight. (very few commits to this repo changes python files).

The `LICENSE` file should be excluded from formatting checks.

And if there is a way to add fewer files to the root of the repository, I would prefer that.

**Comment by arvidn** [comment]
I tried this locally (checking out your branch):

```
pip3 install pre-commit
pre-commit install
pre-commit-run
```

Everything looked OK until `pre-commit run`, which skipped all tests:

```
Trim Trailing Whitespace.............................(no files to check)Skipped
Check Yaml...........................................(no files to check)Skipped
Check for case conflicts.............................(no files to check)Skipped
Check that executables have shebangs.................(no files to check)Skipped
Check Xml............................................(no files to check)Skipped
Debug Statements (Python)............................(no files to check)Skipped
Check for broken symlinks............................(no files to check)Skipped
rst directives end with two colons...................(no files to check)Skipped
rst ``inline code`` next to normal text..............(no files to check)Skipped
flake8...............................................(no files to check)Skipped
```

That's not what it's supposed to do, is it?

**Comment by arvidn** [comment]
or is it only checking files that have changes compared to the top commit? that would make sense I suppose.

**Comment by arvidn** [comment]
I see, it *only* runs on staged files. it specifically stashes unstaged changes while running the checks, and then restores them

**Review by arvidn — APPROVED** [review]

**Comment by AllSeeingEyeTolledEweSew** [comment]
FYI `pre-commit run -a` runs all checks on all files. This is what CI runs.

**Pull request merged**